### ISSUE TITLE
How to Evaluate on New Data

### ISSUE THREAD

**Issue by acp19tag** [kris927b/SkillSpan#2]
Hi, 

Is there a simple way of evaluating a trained model on a new, unseen dataset? 

It looks like `scripts/run.evaluate.sh` needs to look at files in `logs/Skills/etc` specific to the new dataset which are only created during training. 

To clarify, the model I want to evaluate was trained on the supplied data in this repository, but I want to evaluate its performance on my own annotated data!

**Comment by jjzha** [comment]
Hi Tom,

I'd recommend running a similar command indicated in `sctripts/run.predict.sh`. You use a specific model and your dataset.

In a zero-shot setting, note that it might be hard to evaluate one-to-one with your dataset, as it could contain a different label space. E.g., we make use of B/I-Skill, B/I-Knowledge etc. You can do some post-processing on the predictions or your dataset to fit your need. 

Otherwise, I'd recommend fine-tuning `jobbert-base-cased` or similar on your dataset, i.e., creating a `config` file for your task.

Hope this helps!

**Comment by acp19tag** [comment]
Hi, 

Thanks for your quick response!

I'm having some issues doing that, and so I tried running `scripts/run.predict.sh` on the original data following the instructions in the readme.
I also amended this file to reflect the filename changes (`corpus_*` -> `skillspan_*`).

Here are the parameters I want to run:
`$MODEL = spanbert`
`$EXPERIMENT = skills`
`$SET = skills`
 
So I ran:
`bash scripts/run.predict.sh spanbert skills skills > out.txt 2>&1`

This does not appear to run correctly due to 'unrecognized arguments' but I can't work out what the exact issue is. 

I've attached the output log for reference.
[out.txt](https://github.com/kris927b/SkillSpan/files/9131290/out.txt)

**Comment by jjzha** [comment]
Perhaps try to run the `python` command as a standalone, without the bash file. Here, I think your command should be as follows:

`python3 machamp/predict.py [archive_file/model_file] [input_file] [pred_file]` is the structure, below an example:

`python3 machamp/predict.py logs/spanbert/*/model.tar.gz data/skills.conll output.out`

How about this?

**Comment by acp19tag** [comment]
I ran that with a couple of amendments:

- There is no `logs/spanbert/`; the `model_file` directory format is `logs/skill.[model name].[task].[number]/` 
- There is no `data/skills.conll`; the `input_file` name format is `data/Skills/skillspan_[house/tech]_[train/dev/test].conll`

So the command I ran is:

`python3 machamp/predict.py logs/skill.spanbert.skills.1/*/model.tar.gz data/Skills/skillspan_house_test.conll output.out`

I am still getting the same error: 

`usage: predict.py [-h] [--dataset DATASET] [--device DEVICE]
                  [--batch_size BATCH_SIZE] [--raw_text]
                  archive input_file pred_file
predict.py: error: unrecognized arguments: logs/skill.spanbert.skills.1/2022.07.13_16.31.04/model.tar.gz data/Skills/skillspan_house_test.conll output.out`

Here is the output of the command `ls logs/skill.spanbert.skills.1/`:

`2022.07.08_11.53.09  2022.07.08_12.05.41  2022.07.11_10.51.36  2022.07.13_16.31.04`

Therefore the directory `logs/skill.spanbert.skills.1/2022.07.13_16.31.04/` does exist, and does in fact contain the same files as the others - including the `model.tar.gz`.

Any ideas how to solve?

**Comment by jjzha** [comment]
Interesting, I'm not sure. 

Could you also try to give the exact path for the model? So instead of `*` --> `logs/skill.spanbert.skills.1/2022.07.13_16.31.04/model.tar.gz`?

Additionally, I always make the mistake that the paths are incorrect 😛 . Could you double-check where you invoke the `predict.py` from?

**Comment by acp19tag** [comment]
That did it - thanks! 

The command I ran was:

`python3 machamp/predict.py logs/skill.spanbert.skills.1/2022.07.13_16.31.04/model.tar.gz data/Skills/skillspan_house_test.conll output.out --dataset house`

The final line of the output was:

`{'.run/.sum': 0.5198675496688243, '.run/Skills/span_f1': 0.5198675496688243}`
